PartDesign: Transform tools

PaddleStroke opened the pull request:
Fix https://github.com/FreeCAD/FreeCAD/issues/19026

Video presentation : https://youtu.be/cwsvX4YxFsY

Funded by https://github.com/FreeCAD/FPA-grant-proposals/issues/38

Goals addressed : 

- Add spacing patterns & spacing customize (19026) to PD linear/polar patterns objects
- Add UI to easily use spacing pattern
- Refactor TaskLinearPatternParameters & TaskPolarPatternParameters so that there is only TaskPatternParameters.
- Externalize the pattern parameters from TaskPatternParameters and put them into PatternParametersWidget

Video presenting the changes: https://www.youtube.com/watch?v=cwsvX4YxFsY

NOTE: This PR includes the removal of transformed feature message taskbox : 
![image](https://github.com/user-attachments/assets/eacf8110-ebd1-41ab-9b34-0f62effd8c57)

It can be reversed if the @FreeCAD/design-working-group  deems it better without.

hyarion commented:
This PR is rather big (6k+ lines of change)
You mention that it includes other things as well. Would it be possible to split that up in separate commits so we can work with smaller chunks?

There's also a lot of traces of ai generated code and comments. Did you reviewed the code before creating a PR?

luzpaz commented:
`Warning: src/Mod/PartDesign/App/FeatureLinearPattern.cpp:366: occurences ==> occurrences`

Reqrefusion commented:
> 6k lines but it's because there are 2 icons of 2k each. It's just a 2k PR. And a lot of it is refactoring code:
> 
>     * Move ComboLink combobox to Gui.
> 
>     * TaskLinearPatternParameters & TaskPolarPatternParameters become TaskPatternParameters
> 
>     * Creation of PatternParametersWidget
> 
> 
> There isn't so much new code really.
> 
> I can split up small chunks, like the removal of the message box and the feature selection. But there will still be one main chunk.

Yes, you are right, the icons are really big, so they put at least 4 thousand lines on them. 

There is also probably a change of less than 2 thousand lines. The operation "TaskLinearPatternParameters & TaskPolarPatternParameters become TaskPatternParameters" counts the old lines from the new line that are not fully detected by Git. 

If they're all part of a feature, I think it's only right that they're a single PR. I remember that we had problems with this in the past and decided to do so at the developer meeting.

Also, thank you for starting your grant work right away.

kadet1090 reviewed (changes requested):
Few comments here and there. I only saw that you actually moved the ComboLinks not wrote it - it would be however good to refactor it if it is going to be more generic utility. In some comments I pointed you to alternative, more idiomatic ways of doing some things - while it is not strictly necessary to apply them it would be good to keep the codebase idiomatic and consistent.

maxwxyz commented on src/Mod/Part/Gui/PatternParametersWidget.ui:
```diff
@@ -0,0 +1,233 @@
+           </sizepolicy>
+          </property>
+          <property name="toolTip">
+           <string>Reverse the direction of the pattern.</string>
```
```suggestion
           <string>Reverses the direction of the pattern</string>

```

maxwxyz commented on src/Mod/Part/Gui/PatternParametersWidget.ui:
```diff
@@ -0,0 +1,233 @@
+            </sizepolicy>
+           </property>
+           <property name="toolTip">
+            <string>Add spacing to create spacing patterns.</string>
```
```suggestion
            <string>Adds spacing to create spacing patterns</string>

```
@ryankembrey any better suggestion?

maxwxyz commented on src/Mod/PartDesign/Gui/TaskPatternParameters.cpp:
```diff
@@ -0,0 +1,379 @@
+    App::DocumentObject* selObj = nullptr;
+    getReferencedSelection(patternObj, msg, selObj, directions);
+    if (!selObj) {
+        Base::Console().warning(tr("Invalid selection. Please select an edge, planar face, or datum line.").toStdString().c_str());
```
```suggestion
        Base::Console().warning(tr("Invalid selection. Select an edge, planar face, or datum line.").toStdString().c_str());

```

PaddleStroke commented on src/Gui/ComboLinks.h:
```diff
@@ -0,0 +1,172 @@
+        ~ComboLinks();
+
+        // Disable copy/move semantics for simplicity with pointer management
+        ComboLinks(const ComboLinks&) = delete;
+        ComboLinks& operator=(const ComboLinks&) = delete;
+        ComboLinks(ComboLinks&&) = delete;
+        ComboLinks& operator=(ComboLinks&&) = delete;
```
So FC_DISABLE_COPY_MOVE replace the 5 lines correct?

kadet1090 replied:
Yep.

PaddleStroke commented on src/Mod/Part/Gui/PatternParametersWidget.cpp:
```diff
@@ -0,0 +1,478 @@
+    if (blockUpdate || !m_feature) { // Need properties to be bound
+        return;
+    }
+    blockUpdate = true;
```
So instead of doing 
```
    blockUpdate = true;
    //blabla
    blockUpdate = false;
```
We do 
```
    Base::StateLocker locker(&blockUpdate, true);
    //blabla
```
?

kadet1090 replied:
ah yes, it takes reference not pointer - exactly.

PaddleStroke commented on src/Mod/Part/Gui/PatternParametersWidget.cpp:
```diff
@@ -0,0 +1,478 @@
+    // Note: Connections for dynamic rows are done in addSpacingRow()
+}
+
+void PatternParametersWidget::bindProperties(App::PropertyLinkSub& directionProp,
+    App::PropertyBool& reversedProp,
+    App::PropertyEnumeration& modeProp,
+    App::PropertyQuantity& lengthProp,
+    App::PropertyQuantity& offsetProp,
+    App::PropertyFloatList& spacingPatternProp,
+    App::PropertyIntegerConstraint& occurrencesProp,
+    App::DocumentObject* feature)
+{
+    // Store pointers to the properties
+    m_directionProp = &directionProp;
+    m_reversedProp = &reversedProp;
+    m_modeProp = &modeProp;
+    m_extentProp = &lengthProp;
+    m_spacingProp = &offsetProp;
+    m_spacingPatternProp = &spacingPatternProp;
+    m_occurrencesProp = &occurrencesProp;
```
It was to keep the call to this function clean : 
```
        parametersWidget2->bindProperties(linear->Direction2,
            linear->Reversed2,
            linear->Mode2,
            linear->Length2,
            linear->Offset2,
            linear->SpacingPattern2,
            linear->Occurrences2,
            linear);
```

Now it has to be 
```
        parametersWidget2->bindProperties(&linear->Direction2,
            &linear->Reversed2,
            &linear->Mode2,
            &linear->Length2,
            &linear->Offset2,
            &linear->SpacingPattern2,
            &linear->Occurrences2,
            linear);
```

Imo a regression in clarity

PaddleStroke replied:
Please knowing that confirm if you still want this change.

kadet1090 replied:
Yes, 100% sure, you pass it as pointer anyway but in the background. There is a reason why you need & operator to be explicit as this operation can have serious consequences. You are hiding information about that information from the call-site. 

Is it less readable? Maybe, but at least it is clear that this function receives pointer and that you need to be extra careful about lifetimes - this should catch your eye and make you more aware about what you are passing.

PaddleStroke commented on src/Mod/Part/Gui/PatternParametersWidget.cpp:
```diff
@@ -0,0 +1,478 @@
+    ui->spinExtent->bind(*m_extentProp);
+    Base::Unit unit = type == PatternType::Linear ? Base::Unit::Length : Base::Unit::Angle;
+
+    ui->spinExtent->blockSignals(true);
+    ui->spinExtent->setUnit(unit);
+    ui->spinExtent->blockSignals(false);
```
Imo no. This syntax with the block is harder to understand for a c++ noob. The other is very explicit in what it does.

PaddleStroke commented:
@kadet1090 comments all fixed and code refactored for duplicity

kadet1090 commented:
Thanks! I should have some time today evening to test it and do second round of review.

PaddleStroke commented on src/Mod/Part/Gui/PatternParametersWidget.cpp:
```diff
@@ -0,0 +1,471 @@
+
+void PatternParametersWidget::onDirectionChanged(int /*index*/)
+{
+    if (blockUpdate || !m_directionProp) return;
```
Yes that was moved code I usually put braces and replace when I see none.

PaddleStroke commented on src/Mod/PartDesign/Gui/TaskTransformedParameters.cpp:
```diff
@@ -125,38 +111,132 @@ void TaskTransformedParameters::setupUI()
             item->setData(Qt::UserRole, QString::fromLatin1(obj->getNameInDocument()));
+            item->setFlags(item->flags() | Qt::ItemIsUserCheckable);
+
+            Gui::ViewProvider* vp = Application::Instance->getViewProvider(obj);
+            if (vp) {
+                QIcon icon = vp->getIcon();  // Get the icon from the ViewProvider
+                if (!icon.isNull()) {
```
This syntax is confusing imo not an improvement.

kadet1090 replied:
For the second if - I'd not argue - matter of taste imho without objective benefits.

The first however is constructed in a way that the `vp` symbol is available only if it not a nullptr. It physically prevents you from using the unchecked vp because code will simply not compile as the `vp` scope is the true branch of if. So there is objective benefit for using that syntax.

In that case it does not matter that much so feel free to leave it as is.

PaddleStroke commented on src/Mod/PartDesign/Gui/TaskTransformedParameters.cpp:
```diff
@@ -125,38 +111,132 @@ void TaskTransformedParameters::setupUI()
+    }
+
+    std::vector<App::DocumentObject*> originals = pcTransformed->Originals.getValues();
+    bool isChecked = (item->checkState() == Qt::Checked);
+    auto it = std::ranges::find(originals, changedObj);
+    bool changed = false;
```
This is really nitpicking.....

kadet1090 replied:
May seem like nitpicking, but small details like that really matter for reading and writing code later. Keep things consistent at least locally and then there will be no need for comments like that.

kadet1090 commented:
After consulting with DWG we believe that we should stay with the previous feature selection mechanism. While the current one is not ideal (to say the least) and proposal targets some of the pain points it does introduce other problems (like ones that I mentioned in previous comment) and it should be addressed as part of the bigger issue: https://github.com/FreeCAD/FreeCAD/issues/14432.

PaddleStroke commented:
Got it. I'll update when I can.

PaddleStroke commented on src/Mod/PartDesign/App/FeaturePolarPattern.cpp:
```diff
@@ -239,8 +259,36 @@ void PolarPattern::onChanged(const App::Property* prop)
+}
+
+
+void PolarPattern::updateSpacings()
```
FeaturePolar and FeatureLinear should have a base class and those moved to this. But this requires again more refactoring. Perhaps this can be done later

PaddleStroke commented:
> UX issue: Why spacing is under the "Axis" group box?
> UX / text isue: PolarPattern in the header should be written as Polar Pattern

Can you specify what you mean by those? I cannot understand there is no "Axis" groupbox and the header show correct title on my side
<img width="557" height="267" alt="image" src="https://github.com/user-attachments/assets/6151257c-aa92-4c7a-8ce9-2382fa3664e3" />

PaddleStroke commented:
> I think that you by accident removed code that calculated proper value of spacing / extent if switching from one mode to another - which was nice quality of feature.

And this I find this working on my side. But perhaps we don't mean the same thing. Can you specify?
If I put extent = 100, occurence = 5, then change to spacing, the spacing updates to 25 which seems correct to me.

kadet1090 commented:
> If I put extent = 100, occurence = 5, then change to spacing, the spacing updates to 25 which seems correct to me.
I retested it - either I tested something wrong or you fixed it doing something else. Works now.

> Can you specify what you mean by those? I cannot understand there is no "Axis" groupbox and the header show correct title on my side 
<img width="511" height="261" alt="image" src="https://github.com/user-attachments/assets/a3d8c29b-a7dd-435f-97e1-8784ddde00b1" />

The UI reads like the Length and Occurences is property of Axis - and they are not, they are properties of the pattern.

For the Linear pattern it kind of works because "Direction" is simply a better label with more general meaning and it is reinforced by having two separate directions so the grouping makes sense. For "Axis" it does not work too well as the term is more specific and you don't have second axis so grouping is not necessary.

Here I'd remove the group box altogether and put the "Axis" label on the left of the selector. That would make the UI consistent. If more than one axis would be introduced I'd keep the "Direction" label for both Polar and Linear pattern and only switch the label of the selected reference. For the linear pattern it could simply be "Reference" and for polar it would be "Axis".

Furthermore, for Linear Pattern the second direction is always visible - it should be conditionally available with some kind of checkbox as with the current state it really suggests that it is not possible to create pattern in one direction only. 

Lastly in accordance to the text guidelines (https://github.com/FreeCAD/FreeCAD-Enhancement-Proposals/blob/master/FEPs/FEP-0007-consistent-language/README.md) you should avoid using numbers and instead spell them out - so "Direction 2" should become "Second direction".

> For the + sign I don't see how to change it so that it makes sense.

<img width="372" height="327" alt="image" src="https://github.com/user-attachments/assets/c5cdb5f3-fa23-434c-a845-8baa85f57f71" />

I've added a label to the button also as buttons really should be labeled - the plus sign is not clear enough imo - but not a hill that I'll die on. Ideally you should also be able to remove the first entry and assign expressions to every of them but it would be probably problematic. 

That UI is also not perfect and probably should be unified with something similar to the object picker that I suggested some time ago: 

<img width="369" height="447" alt="image" src="https://github.com/user-attachments/assets/838866f2-fecb-43ac-b87d-76c9c8f86992" />

It would allow user to select multiple offsets and delete them at the same time without needing to click the "-" button multiple times. hat however would be too massive change for now and should be done as part of unification effort for fields with multiple entries.

PaddleStroke commented:
ok I see the axis groupbox now, I was looking at linear. Yes this name does not work.

It is however difficult to remove the groupbox since it's part of PatternParametersWidget. So it's going to be a big pain to change.

Perhaps we can just call it Direction rather than Axis.

pierreporte commented:
Yes I’ve seen the video at the time and I didn’t find that part of the feature very good from start. New users are going to be completely lost though. It adds a lot of complexity compared to cascading patterns.

Maybe it would be better to have an editable list of all the gaps, folded by default. So if you have 10 occurrences, you get a 9-row list, in which you can edit exactly the needed gaps. That list would be deleted once interactive controls are able to replace it.

pierreporte commented:
Also I think that the default mode should be spacing for linear pattern, and extent for polar pattern.

pierreporte commented:
Actually the multi-transform feature is already there for composition, and you modified it to be consistent with the individual feature. So there is really no need to add the spacing pattern. It’s even useless within multi-transform as it duplicates the very purpose of the feature. So really, I think that the best you could do is to replace that by the custom gaps until the interactive controls and OVPs are implemented.

One thing I noticed in multi-transform is that you need to double click on the feature to edit it. It should just require a single-click selection. Not sure about the OK button also. In general the UI for stack of transformations could be better.

PaddleStroke commented:
Indeed it appears to be redundantwith multitransform as you can achieve the same with a pattern of patterns.

Well then I can remove that feature I guess. @FreeCAD/design-working-group everyone agrees it's best to revert this feature?

PaddleStroke commented:
@pierreporte I was about to revert this, and found an advantage on this compared to multitransform.
With the 'spacing patterns' you can make a pattern and you can override any gap at any position. Which you cannot do with multitransform. With multitransform you can't.

Take this example for instance : 
1 2 3 1 2 3 1 2 3 1 2 3

With multi transform you can do it by : 
- a first linear pattern with spacing = 1 then you override the second and third gap to 2 and 3.
- a second linear pattern of the first.

But for this example : 
1 **5** 3 1 2 3 1 2 3 1 2 3

Here you can't do it with multi transform.

Thoughts?

PaddleStroke commented:
Ok so I suggest that I remove only the UI (ie I remove the + button) but leave the property. Or perhaps add a parameter that hide the button by default. So if some users wants to use the feature he can add the parameter.

This way when we have OVP implemented we can re-asses the situation. And meanwhile this feature can be kept in AstoCAD so that we get some user feedback on it.

Does that sounds good @kadet1090 ?

kadet1090 commented:
Yes, let's proceed with that as it is the safest option, that would enable us to merge the PR faster. I'd only ask to add some `(Experimental)` text to the tooltip so users are informed that the feature may change in the future.

Possibly we could make a better UI for it to better express how it works (for example instead of calling it spacing, name it pattern and arrange it a bit differently - just thinking at loud).

pierreporte commented:
@kadet1090 you mean a tooltip in the property view? I haven't realized they exist.

PaddleStroke commented:
Actually there are currently no tooltips to any property in patterns. So I'll add those.

kadet1090 commented:
> @kadet1090 it's ready

Thanks! I'll review and test it today and if there won't be big issues (I don't expect any) I'll merge today.

PaddleStroke commented:
Actually no that's not the reason. Now I remember : I removed it because it was redundant : 
<img width="938" height="971" alt="image" src="https://github.com/user-attachments/assets/43c24e88-d140-48ab-aef5-a1556ad718c7" />

- you have a warning '3 transformed shapes do not intersect...' that gets printed all over the place.
- when I initially made this feature, the non intersecting shapes were shown in red. Clearly showing that some shapes were not going to be added.

So I thought that warning I removed was very redundant.

FEA-eng commented:
> @FEA-eng see [#22389 (comment)](https://github.com/FreeCAD/FreeCAD/pull/22389#issuecomment-3204265330) and previous messages, the button was hidden by default

Ok, thanks. So it should be fully available in 1.2 with OVP for it.

PaddleStroke commented:
There're 2 things : 
- Spacing override : This is the Spacings property of the pattern object. It let you override specific occurence of spacing. It has no UI currently. There is a open PR to add OVP for it.
- Spacing patterns : This is the SpacingPattern property of pattern object. This has a UI, the + button, but it was disabled by default because people thought it was overkill and redundant with multitransform.
